DM-32772: Update for v23.0.0 release (#195)
lsst/pipelines_lsst_io: releases/v23_0_0.rst

The notes below highlight significant technical changes to the Science Pipelines codebase in this release.
For a complete list of changes made, see :doc:`tickets/v23_0_0`.

The `Characterization Metric Report (DMTR-351) <https://dmtr-351.lsst.io/>`_ describes the scientific performance of this release in terms of scientific performance metrics.

If you have questions or comments about this release, visit our `community forum <https://community.lsst.org>`_ for advice.

- :ref:`release-v23-0-0-functionality`
- :ref:`release-v23-0-0-interface`
- :ref:`release-v23-0-0-pending-deprecations`
- :ref:`release-v23-0-0-deprecations`

*See also:*

- :doc:`Installation instructions </install/index>`
- :doc:`Known issues </known-issues>`
- `Doxygen Documentation`__

__ http://doxygen.lsst.codes/stack/doxygen/xlink_master_2021_09_30_20.02.36/

.. _release-v23-0-0-functionality:

Major New Features
==================

- :ref:`release-v23-0-0-gen3`
- :ref:`release-v23-0-0-meas_extensions_gaap`
- :ref:`release-v23-0-0-trailed`
- :ref:`release-v23-0-0-cp_verify`
- :ref:`release-v23-0-0-ci_imsim`

.. _release-v23-0-0-gen3:

Generation 3 Middleware
-----------------------

> Generation 3 Middleware is being used in production for the DP0.2 data release. This release is that last with functional generation 2 middleware. See :ref:`release-v23-0-0-deprecate-gen2`. Some highlights include:

Reviewer: ```suggestion
Generation 3 Middleware is being used for the DP0.2 data release production. This is the last release with functional generation 2 middleware. See :ref:`release-v23-0-0-deprecate-gen2`. Some highlights include:
```

* Added a new facility for creating lightweight butlers that pre-fills a local SQLite registry. This "execution butler" can allow a pipeline to be executed without talking to the main registry, reducing contention for connections to shared registries. :jirab:`DM-28646`
* Log diagnostic information when QuantumGraphs are empty because the initial query yielded no results. At present, these diagnostics only cover missing input datasets, which is a common way to get an empty QuantumGraph, but not the only way. :jirab:`DM-31583`
* Add ability to cache datasets locally when using a remote file store. This can significantly improve performance when retrieving components from a dataset. :jirab:`DM-13365`

> For more information, see detailed notes on changes to the following middleware packages on this release: `daf_butler <https://pipelines.lsst.io/v/daily/modules/lsst.daf.butler/CHANGES.html>`_, `pipe_base <https://pipelines.lsst.io/v/daily/modules/lsst.pipe.base/CHANGES.html>`_, `ctrl_mpexec <https://pipelines.lsst.io/v/daily/modules/lsst.ctrl.mpexec/CHANGES.html>`_, and `ctrl_bps <https://pipelines.lsst.io/v/daily/modules/lsst.ctrl.bps/CHANGES.html>`_.

Reviewer: These are external links to the daily docs, which is a no-go. These should work as internal links:

```suggestion
For more information, see detailed notes on changes to the following middleware packages on this release:

- :doc:`lsst.daf.butler </modules/lsst.daf.base/CHANGES>`_,
- :doc:`lsst.pipe.base </modules/lsst.pipe.base/CHANGES>`_,
- :doc:`lsst.ctrl.mpexec </modules/lsst.ctrl.mpexec/CHANGES>`_,
- :doc:`lsst.ctrl.bps </modules/lsst.ctrl.bps/CHANGES>`_.
```

Author: I have to undo this change.

The 23.0.0 middleware package release notes aren't in 23.0.0 (because they were compiled after the final release candidate rc5), so those internal links don't exist until we build 23.0.1.  e.g.: https://lsst.ncsa.illinois.edu/~yusra/23.0.0/modules/lsst.ctrl.mpexec/index.html#changes

https://lsstc.slack.com/archives/C2JPMCF5X/p1639601968158000?thread_ts=1639601844.157900&cid=C2JPMCF5X

Recommendations are welcome. I could just not link anything and say that more details are on the way? I could leave the dead links which will magically get switched on for v23.0.1.

Author: Forgot to mention: I'll use internal links on the version that gets merged to main.

OK how about this for v23.0.x: Added a TODO DM-33028 for the internal links to be backported. Made the external links go to github, so that it'll be super obvious that they're not looking at the 23.0.0 doc build anymore...

Reviewer: Yeah, pragmatism for now and patching it is fine if that's what you want. As you can imagine, we can't do this style of linking in general because these v23 release notes will someday be linking to the v28 release notes for butler :)

We definitely need to formalize the change log management of middleware and put it into the Science Pipelines release procedure; I'm not thrilled with how the middleware team kind of forked the release notes process unilaterally.

Reviewer 2: Middleware did not do anything that required any changes to process so I'm not sure why this is causing upset. I wanted to keep track of changes as we made them and we adopted a scheme that made that easy. The global release notes were not affected. I made the middleware release notes for v23 after rc5 so they don't appear in the butler docs. If there was an rc6 then they would appear but that doesn't affect anything here. If you link generically to CHANGES then there are sections for each release.

Author: > We definitely need to formalize the change log management of middleware and put it into the Science Pipelines release procedure; 

We prob should have done one more rc.  This is the first time we've had a long-lived release branch tied to a data release, with more than a couple backported tickets and release candidates. So, we're still working out the final procedure in this new realm. 

We were going to try using towncrier in more packages next cycle. It was kind of nice to select some highlights to copy and paste.

Gen2 infrastructure code (within e.g., :ref:`pipe_tasks <lsst.pipe.tasks>`, :ref:`pipe_base <lsst.pipe.base>`, :ref:`obs_base <lsst.obs.base>`, `daf_persistence <https://github.com:lsst/daf_persistence>`_, and obs packages) will no longer be maintained after January 1 2022 and may be removed at any point afterward.
The CI package ``ci_hsc_gen2`` will continue to be run, and the tasks it checks will be maintained until we remove the Gen2 infrastructure code in 2022.
Following this release, we will remove Gen2 infrastructure code and the commandline-task entry points, ``runDataRef`` to the Tasks.

.. _release-v23-0-0-deprecate-decamRawIngestTask:

Deprecated DecamRawIngestTask and MegaPrimeRawIngestTask
-----------------------------------------------------------------
MegaPrime and DECam no longer require a specialist Gen3 ingest task.
Please use the default `~lsst.obs.base.RawIngestTask`.
Both `~lsst.obs.decam.DecamRawIngestTask` and `~lsst.obs.cfht.MegaPrimeRawIngestTask` will be removed after v23.

.. _release-v23-0-0-deprecate-Filter:

Removal of ``lsst.afw.Filter``
------------------------------

> The `lsst.afw.Filter` class has been replaced with :ref:`filterLabel <release-v22-0-0-filterLabel>`, is deprecated, and will be removed after this release.

Reviewer: - Check the API name (not that it matters anymore). Is it actually `lsst.afw.image.Filter`?
- The link to `lsst.afw.Filter` doesn't work; you might want to use double backticks because I don't think that API is documented anyways (?)

Method names that contain ``Filter``, such as ``getFilter``, have been replaced e.g., with ``.getFilterLabel``.
:jirab:`RFC-730`

.. _release-v23-0-0-deprecate-validate-drp:

Removal of validate_drp
-----------------------

> The algorithms implemented in :ref:`validate_drp <lsst.validate.drp>` were ported as-is to run in :ref:`faro <lsst.faro.base>`.

Reviewer: ```suggestion
The algorithms implemented in :ref:`lsst.validate.drp` were ported as-is to run in :ref:`lsst.faro`.
```

All future development of scientific performance metrics will be carried out in faro.

> :ref:`validate_drp <lsst.validate.drp>` will be removed after this release.

Reviewer: ```suggestion
The ``validate_drp`` EUPS package will be removed after this release.
```
